### PR TITLE
Add execution summary framework for non-batch executors

### PR DESCRIPTION
**Pull request by breeswish**
Signed-off-by: Breezewish <breezewish@pingcap.com>

## What have you changed? (mandatory)

Extracted from https://github.com/tikv/tikv/pull/4598

This PR adds support of the execution summary for non-batch executors.

## What are the type of the changes? (mandatory)

- Engineering (engineering change which doesn't change any feature or fix any issue)

## How has this PR been tested? (mandatory)

See https://github.com/tikv/tikv/pull/4598

### CODE REVIEW

File: src/coprocessor/dag/builder.rs [hunk 1]
```diff
@@ -341,7 +355,7 @@ impl DAGBuilder {
             let build_batch_result =
                 super::builder::DAGBuilder::check_build_batch(req.get_executors());
             if let Err(e) = build_batch_result {
-                debug!("Coprocessor request cannot be batched"; "reason" => %e);
+                info!("Coprocessor request cannot be batched"; "reason" => %e);
```
Comment (breeswish) [inline src/coprocessor/dag/builder.rs]: Changed intentionally.

File: src/coprocessor/dag/handler.rs [hunk 2]
```diff
@@ -22,12 +27,16 @@ impl DAGRequestHandler {
         executor: Box<dyn Executor + Send>,
         output_offsets: Vec<u32>,
         batch_row_limit: usize,
+        number_of_executors: usize,
+        collect_execution_summary: bool,
```
Comment (rleungx) [inline src/coprocessor/dag/handler.rs]: perfer to use `collect_execution_summary` to keep consistent with batch

File: src/coprocessor/dag/executor/limit.rs [hunk 3]
```diff
@@ -3,6 +3,8 @@
 use tipb::executor::Limit;
 
 use super::ExecutorMetrics;
+
```
Comment (rleungx) [inline src/coprocessor/dag/executor/limit.rs]: remove this line?

**Pull request merged**